ABI encodeFunctionCall

Hello,
Could it be possible to get the web3 `encodeFunctionCall`?

=> https://web3js.readthedocs.io/en/v1.5.2/web3-eth-abi.html#encodefunctioncall

I tried to make something but I did not completely get how to format dynamic types.
It would be very helpful if you can provide an out-of-the-box function that does this.

Hey @ferranbt, I made a unit test to compare what the smart contract is returning vs what type encode for `uint256[]` returns:

```go
func TestEncodeWithType(t *testing.T) {
	utests := []struct {
		scenario string
		typ      string
		val      interface{}
		expected string
	}{
		{
			scenario: "empty uint256[]",
			typ:      "uint256[]",
			val:      []uint{},
			expected: "0x00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000000",
		},
		{
			scenario: "uint256[] with 1 element",
			typ:      "uint256[]",
			val:      []uint{1},
			expected: "0x000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000001",
		},
	}

	for _, u := range utests {
		t.Run(u.scenario, func(t *testing.T) {
			typ, err := abi.NewType(u.typ)
			require.NoError(t, err)

			data, err := typ.Encode(u.val)
			require.NoError(t, err)
			require.Equal(t, u.expected, fmt.Sprintf("0x%x", data))
		})
	}
}
```

Output:

```sh
▶ go test
--- FAIL: TestEncodeWithType (0.00s)
    --- FAIL: TestEncodeWithType/empty_uint256[] (0.00s)
        contract_test.go:93:
            	Error Trace:	contract_test.go:93
            	Error:      	Not equal:
            	            	expected: "0x00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000000"
            	            	actual  : "0x0000000000000000000000000000000000000000000000000000000000000000"

            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1 +1 @@
            	            	-0x00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000000
            	            	+0x0000000000000000000000000000000000000000000000000000000000000000
            	Test:       	TestEncodeWithType/empty_uint256[]
    --- FAIL: TestEncodeWithType/uint256[]_with_1_element (0.00s)
        contract_test.go:93:
            	Error Trace:	contract_test.go:93
            	Error:      	Not equal:
            	            	expected: "0x000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000001"
            	            	actual  : "0x00000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000001"

            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1 +1 @@
            	            	-0x000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000001
            	            	+0x00000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000001
            	Test:       	TestEncodeWithType/uint256[]_with_1_element
FAIL
exit status 1
FAIL	github.com/maxence-charriere/nftile/pkg/eth	0.295s
```

Look like the one returned by the smart contract has some header data.

Note that I was able to decode the smart contract function return with [web3js decodeParameter](https://web3js.readthedocs.io/en/v1.5.2/web3-eth-abi.html#decodeparameter) and retrieve the expected array.

The solidity version I use is `0.8.4`.

I also made a test to try encoding from a method and noticed it was failing:
```go
func TestEncodeFromMethod(t *testing.T) {
	utests := []struct {
		scenario string
		val      []uint
	}{
		{
			scenario: "encode empty slice",
			val:      []uint{},
		},
		{
			scenario: "encode slice with 1 element",
			val:      []uint{1},
		},
	}

	for _, u := range utests {
		t.Run(u.scenario, func(t *testing.T) {
			f := "function getTileIdsByPublisher(address publisher) view returns (uint256[])"
			m, err := abi.NewMethod(f)
			require.NoError(t, err)

			_, err = m.Outputs.Encode([]uint{1})
			require.NoError(t, err)
		})
	}
}
```

Output:
```sh
--- FAIL: TestEncodeFromMethod (0.00s)
    --- FAIL: TestEncodeFromMethod/encode_empty_slice (0.00s)
        contract_test.go:120:
            	Error Trace:	contract_test.go:120
            	Error:      	Received unexpected error:
            	            	failed to encode uint as Slice
            	Test:       	TestEncodeFromMethod/encode_empty_slice
    --- FAIL: TestEncodeFromMethod/encode_slice_with_1_element (0.00s)
        contract_test.go:120:
            	Error Trace:	contract_test.go:120
            	Error:      	Received unexpected error:
            	            	failed to encode uint as Slice
            	Test:       	TestEncodeFromMethod/encode_slice_with_1_element
FAIL
exit status 1
FAIL	github.com/maxence-charriere/nftile/pkg/eth	2.366s
```

--- a/abi/abi.go
+++ b/abi/abi.go
@@ -21,6 +21,11 @@ type ABI struct {
 	Events      map[string]*Event
 }
 
+func (a *ABI) GetMethod(name string) *Method {
+	m := a.Methods[name]
+	return m
+}
+
 func (a *ABI) addEvent(e *Event) {
 	if len(a.Methods) == 0 {
 		a.Events = map[string]*Event{}
@@ -139,6 +144,29 @@ func (m *Method) ID() []byte {
 	dst := k.Sum(nil)[:4]
 	releaseKeccak(k)
 	return dst
+}
+
+// Encode encodes the inputs with this function
+func (m *Method) Encode(args ...interface{}) ([]byte, error) {
+	data, err := Encode(args, m.Inputs)
+	if err != nil {
+		return nil, err
+	}
+	data = append(m.ID(), data...)
+	return data, nil
+}
+
+// Decode decodes the output with this function
+func (m *Method) Decode(data []byte) (map[string]interface{}, error) {
+	if len(data) == 0 {
+		return nil, fmt.Errorf("empty response")
+	}
+	respInterface, err := Decode(m.Outputs, data)
+	if err != nil {
+		return nil, err
+	}
+	resp := respInterface.(map[string]interface{})
+	return resp, nil
 }
 
 func NewMethod(name string) (*Method, error) {

--- a/abi/abi.go
+++ b/abi/abi.go
@@ -147,7 +147,7 @@ func (m *Method) ID() []byte {
 }
 
 // Encode encodes the inputs with this function
-func (m *Method) Encode(args ...interface{}) ([]byte, error) {
+func (m *Method) Encode(args interface{}) ([]byte, error) {
 	data, err := Encode(args, m.Inputs)
 	if err != nil {
 		return nil, err

--- a/abi/encoding_test.go
+++ b/abi/encoding_test.go
@@ -76,8 +76,8 @@ func TestEncoding(t *testing.T) {
 		{
 			"bytes10[]",
 			[][10]byte{
-				[10]byte{0x1, 0x2, 0x3, 0x4, 0x5, 0x6, 0x7, 0x8, 0x9, 0x10},
-				[10]byte{0x1, 0x2, 0x3, 0x4, 0x5, 0x6, 0x7, 0x8, 0x9, 0x10},
+				{0x1, 0x2, 0x3, 0x4, 0x5, 0x6, 0x7, 0x8, 0x9, 0x10},
+				{0x1, 0x2, 0x3, 0x4, 0x5, 0x6, 0x7, 0x8, 0x9, 0x10},
 			},
 		},
 		{
@@ -201,11 +201,11 @@ func TestEncoding(t *testing.T) {
 			// tuple array
 			"tuple(int32 a, int32 b)[2]",
 			[2]map[string]interface{}{
-				map[string]interface{}{
+				{
 					"a": int32(1),
 					"b": int32(2),
 				},
-				map[string]interface{}{
+				{
 					"a": int32(3),
 					"b": int32(4),
 				},
@@ -216,10 +216,10 @@ func TestEncoding(t *testing.T) {
 			// tuple array with dynamic content
 			"tuple(int32[] a)[2]",
 			[2]map[string]interface{}{
-				map[string]interface{}{
+				{
 					"a": []int32{1, 2, 3},
 				},
-				map[string]interface{}{
+				{
 					"a": []int32{4, 5, 6},
 				},
 			},
@@ -228,11 +228,11 @@ func TestEncoding(t *testing.T) {
 			// tuple slice
 			"tuple(int32 a, int32[] b)[]",
 			[]map[string]interface{}{
-				map[string]interface{}{
+				{
 					"a": int32(1),
 					"b": []int32{2, 3},
 				},
-				map[string]interface{}{
+				{
 					"a": int32(4),
 					"b": []int32{5, 6},
 				},
@@ -254,11 +254,11 @@ func TestEncoding(t *testing.T) {
 			map[string]interface{}{
 				"a": [2]uint8{uint8(1), uint8(2)},
 				"b": [2]map[string]interface{}{
-					map[string]interface{}{
+					{
 						"e": uint8(10),
 						"f": uint32(11),
 					},
-					map[string]interface{}{
+					{
 						"e": uint8(20),
 						"f": uint32(21),
 					},
@@ -270,26 +270,26 @@ func TestEncoding(t *testing.T) {
 		{
 			"tuple(uint16 a, uint16 b)[1][]",
 			[][1]map[string]interface{}{
-				[1]map[string]interface{}{
-					map[string]interface{}{
+				{
+					{
 						"a": uint16(1),
 						"b": uint16(2),
 					},
 				},
-				[1]map[string]interface{}{
-					map[string]interface{}{
+				{
+					{
 						"a": uint16(3),
 						"b": uint16(4),
 					},
 				},
-				[1]map[string]interface{}{
-					map[string]interface{}{
+				{
+					{
 						"a": uint16(5),
 						"b": uint16(6),
 					},
 				},
-				[1]map[string]interface{}{
-					map[string]interface{}{
+				{
+					{
 						"a": uint16(7),
 						"b": uint16(8),
 					},
@@ -300,10 +300,10 @@ func TestEncoding(t *testing.T) {
 			"tuple(uint64[][] a, tuple(uint8 a, uint32 b)[1] b, uint64 c)",
 			map[string]interface{}{
 				"a": [][]uint64{
-					[]uint64{3, 4},
+					{3, 4},
 				},
 				"b": [1]map[string]interface{}{
-					map[string]interface{}{
+					{
 						"a": uint8(1),
 						"b": uint32(2),
 					},
@@ -339,11 +339,11 @@ func TestEncodingArguments(t *testing.T) {
 			&ArgumentStr{
 				Type: "tuple",
 				Components: []*ArgumentStr{
-					&ArgumentStr{
+					{
 						Name: "",
 						Type: "int32",
 					},
-					&ArgumentStr{
+					{
 						Name: "",
 						Type: "int32",
 					},
@@ -358,11 +358,11 @@ func TestEncodingArguments(t *testing.T) {
 			&ArgumentStr{
 				Type: "tuple",
 				Components: []*ArgumentStr{
-					&ArgumentStr{
+					{
 						Name: "a",
 						Type: "int32",
 					},
-					&ArgumentStr{
+					{
 						Name: "",
 						Type: "int32",
 					},
@@ -493,19 +493,19 @@ func testTypeWithContract(t *testing.T, server *testutil.TestServer, typ *Type) 
 	tt := method.Inputs
 	val := generateRandomType(tt)
 
-	data, err := Encode(val, tt)
+	data, err := method.Encode(val)
 	if err != nil {
 		return err
 	}
 
 	res, err := server.Call(&web3.CallMsg{
 		To:   &receipt.ContractAddress,
-		Data: append(method.ID(), data...),
+		Data: data,
 	})
 	if err != nil {
 		return err
 	}
-	if res != encodeHex(data) {
+	if res != encodeHex(data[4:]) { // remove funct signature in data
 		return fmt.Errorf("bad")
 	}
 	return nil

--- a/contract/contract.go
+++ b/contract/contract.go
@@ -65,7 +65,7 @@ func (c *Contract) Call(method string, block web3.BlockNumber, args ...interface
 		return nil, fmt.Errorf("method %s not found", method)
 	}
 
-	data, err := m.Encode(args...)
+	data, err := m.Encode(args)
 	if err != nil {
 		return nil, err
 	}

--- a/contract/contract.go
+++ b/contract/contract.go
@@ -60,17 +60,15 @@ func (c *Contract) EstimateGas(method string, args ...interface{}) (uint64, erro
 
 // Call calls a method in the contract
 func (c *Contract) Call(method string, block web3.BlockNumber, args ...interface{}) (map[string]interface{}, error) {
-	m, ok := c.abi.Methods[method]
-	if !ok {
+	m := c.abi.GetMethod(method)
+	if m == nil {
 		return nil, fmt.Errorf("method %s not found", method)
 	}
 
-	// Encode input
-	data, err := abi.Encode(args, m.Inputs)
+	data, err := m.Encode(args...)
 	if err != nil {
 		return nil, err
 	}
-	data = append(m.ID(), data...)
 
 	// Call function
 	msg := &web3.CallMsg{
@@ -91,15 +89,10 @@ func (c *Contract) Call(method string, block web3.BlockNumber, args ...interface
 	if err != nil {
 		return nil, err
 	}
-	if len(raw) == 0 {
-		return nil, fmt.Errorf("empty response")
-	}
-	respInterface, err := abi.Decode(m.Outputs, raw)
+	resp, err := m.Decode(raw)
 	if err != nil {
 		return nil, err
 	}
-
-	resp := respInterface.(map[string]interface{})
 	return resp, nil
 }
 

